src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+/**
+ * Implementation of the C interface functions
+ */
+// system
+#include <cstring>
+#include <sstream>
+#include <type_traits> //for std::underlying_type
+
+// libraries
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+
+// local
+#include <XdgUtils/DesktopEntry/DesktopEntry.h>
+#include <appimage/core/AppImage.h>
+#include "utils/HashLib.h"
+#include "utils/UrlEncoder.h"
+
+#ifdef LIBAPPIMAGE_DESKTOP_INTEGRATION_ENABLED
+
+#include <appimage/desktop_integration/IntegrationManager.h>
+#endif
+
+using namespace appimage::core;
+
+namespace bf = boost::filesystem;
+
+extern "C" {
+
+/* Check if a file is an AppImage. Returns the image type if it is, or -1 if it isn't */
+int appimage_get_type(const char* path, bool verbose) {
+    typedef std::underlying_type<AppImageFormat>::type utype;
+
+    try {
+        AppImage appImage(path);
+        return static_cast<utype>(appImage.getFormat());
+    } catch (const std::runtime_error& err) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    return static_cast<utype>(AppImageFormat::INVALID);
+}
+
+char** appimage_list_files(const char* path) {
+    char** result = nullptr;
+    try {
+        AppImage appImage(path);
+
+        std::vector<std::string> files;
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr)
+            if (!(*itr).empty())
+                files.emplace_back(*itr);
+
+
+        result = static_cast<char**>(malloc(sizeof(char*) * (files.size() + 1)));
+        for (int i = 0; i < files.size(); i++)
+            result[i] = strdup(files[i].c_str());
+
+        result[files.size()] = nullptr;
+
+        return result;
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    // Create empty string list
+    result = static_cast<char**>(malloc(sizeof(char*)));
+    result[0] = nullptr;
+
+    return result;
+}
+
+void appimage_string_list_free(char** list) {
+    for (char** ptr = list; ptr != NULL && *ptr != NULL; ptr++)
+        free(*ptr);
+
+    free(list);
+}
+
+bool
+appimage_read_file_into_buffer_following_symlinks(const char* appimage_file_path, const char* file_path, char** buffer,
+                                                  unsigned long* buf_size) {
+    // Init output params
+    *buffer = nullptr;
```

Comment:
In case of errors, these variables shouldn't be modified IMO.  There's no technical reason to "initialize" those.

Reply:
We have to make sure that no valid values are placed there after the function execution even if the user was the one who put that value there.

Reply:
That's not a responsibility of this method. It shouldn't perform this operation therefore.

Reply:
Those arguments are part of the output of the method, therefore they are totally under its responsibility.

Reply:
The responsibility of the method is not to modify the parameters for the sake of doing it. If there's an error, this should _only_ ever be checked through the return value. However, some lazy people just check whether one of the output params they passed changed. You will change them in any case, without any technical reason. It's bad practice to check only whether the output parameters changed, but it's as well bad practice to change output params if there's an error.

Reply:
It's not about people being lazy it's about they use the software as they are used to and as the software allows it. If those input parameters were moved to a struct and returned as part of the function output they will have to be set to valid values.

Reply:
They are not input parameters, they're output parameters, so your comment doesn't make much sense.

Reply:
Sorry, I mean output.